File: sdk/python/phase41_conformance.py
#!/usr/bin/env python3
"""Executable Phase-41 Python SDK host probe."""

from ucr_sdk.auth import (
    CREDENTIAL_ID_METADATA_KEY,
    CREDENTIAL_SECRET_METADATA_KEY,
    ServiceCredential,
)


def require(condition: bool, message: str) -> None:
    if not condition:
        raise SystemExit(message)


def main() -> None:
    credential_id = bytes((0x00, 0x01, 0xFE, 0xFF))
    secret = bytes((0xDE, 0xAD, 0xBE, 0xEF))
    credential = ServiceCredential(credential_id, secret)
    metadata = credential.metadata()

    require(len(metadata) == 2, "Python SDK emitted unexpected metadata entries")
    require(metadata[0] == (CREDENTIAL_ID_METADATA_KEY, credential_id), "credential id drifted")
    require(metadata[1] == (CREDENTIAL_SECRET_METADATA_KEY, secret), "credential secret drifted")

    rendered = repr(credential)
    require("[REDACTED]" in rendered, "Python credential diagnostics lost redaction")
    require("deadbeef" not in rendered.lower(), "Python diagnostics exposed secret bytes")

    # bytes are immutable and therefore cannot be changed through returned metadata.
    require(metadata[0][1] is credential_id, "Python helper unexpectedly rewrote opaque id bytes")
    require(metadata[1][1] is secret, "Python helper unexpectedly rewrote opaque secret bytes")
    print("UCR_PHASE41_PYTHON_OK")


if __name__ == "__main__":
    main()

File: sdk/typescript/phase41_conformance.ts
import {
  CREDENTIAL_ID_METADATA_KEY,
  CREDENTIAL_SECRET_METADATA_KEY,
  ServiceCredential,
} from "./src/auth.ts";

function requireCondition(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

const credentialId = new Uint8Array([0x00, 0x01, 0xfe, 0xff]);
const secret = new Uint8Array([0xde, 0xad, 0xbe, 0xef]);
const credential = new ServiceCredential(credentialId, secret);
const metadata = credential.binaryMetadata();

requireCondition(metadata.length === 2, "TypeScript SDK emitted unexpected metadata entries");
requireCondition(metadata[0].key === CREDENTIAL_ID_METADATA_KEY, "credential id key drifted");
requireCondition(metadata[1].key === CREDENTIAL_SECRET_METADATA_KEY, "credential secret key drifted");
requireCondition(
  Buffer.from(metadata[0].value).equals(Buffer.from(credentialId)),
  "credential id bytes drifted",
);
requireCondition(
  Buffer.from(metadata[1].value).equals(Buffer.from(secret)),
  "credential secret bytes drifted",
);

metadata[0].value[0] = 0x7f;
requireCondition(
  credential.binaryMetadata()[0].value[0] === credentialId[0],
  "returned TypeScript metadata mutated stored credential bytes",
);
const rendered = credential.toString();
requireCondition(rendered.includes("[REDACTED]"), "TypeScript credential diagnostics lost redaction");
requireCondition(!rendered.toLowerCase().includes("deadbeef"), "TypeScript diagnostics exposed secret bytes");

console.log("UCR_PHASE41_TYPESCRIPT_OK");
